File: src/services/posts/avatar.ts
// src/services/posts/avatar.ts

/**
 * DiceBear SVG generation island — the ONLY avatar logic left client-side.
 *
 * Field parsing and priority resolution (Yappr avatar field → DashPay
 * avatarUrl → deterministic thumbs) live Rust-side
 * (src-tauri/src/social/avatar.rs); this module only turns the resolved
 * { style, seed } into an SVG string, byte-identical to yap.pr and
 * evonext-mobile because all three use @dicebear 9.4.x.
 *
 * (Rust's dicebear-core 11.0.0-rc renders DIFFERENT artwork for the same
 * style+seed — verified empirically 2026-09-11 — which is why generation
 * stays here. It is 100% local: no network call, no identity leak. This
 * replaces the old api.dicebear.com/7.x fallback URL, which leaked
 * identity IDs to a third party on every feed render.)
 */

import { createAvatar } from '@dicebear/core'
import {
    adventurer, adventurerNeutral, avataaars, avataaarsNeutral,
    bigEars, bigEarsNeutral, bigSmile, bottts, botttsNeutral,
    croodles, croodlesNeutral, funEmoji, icons, identicon, initials,
    lorelei, loreleiNeutral, micah, miniavs, notionists, notionistsNeutral,
    openPeeps, personas, pixelArt, pixelArtNeutral, rings, shapes, thumbs,
} from '@dicebear/collection'
import type { ISocialAvatarSource } from '@/types/social'

/** Default style for identity-seeded fallback avatars (Yappr uses thumbs). */
const DEFAULT_STYLE = 'thumbs'

/**
 * Style name → DiceBear collection module.
 * Mirrors Yappr's styleMap so `style` values stored in profile documents
 * render identically across clients.
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
const STYLE_MAP: Record<string, any> = {
    'adventurer': adventurer,
    'adventurer-neutral': adventurerNeutral,
    'avataaars': avataaars,
    'avataaars-neutral': avataaarsNeutral,
    'big-ears': bigEars,
    'big-ears-neutral': bigEarsNeutral,
    'big-smile': bigSmile,
    'bottts': bottts,
    'bottts-neutral': botttsNeutral,
    'croodles': croodles,
    'croodles-neutral': croodlesNeutral,
    'fun-emoji': funEmoji,
    'icons': icons,
    'identicon': identicon,
    'initials': initials,
    'lorelei': lorelei,
    'lorelei-neutral': loreleiNeutral,
    'micah': micah,
    'miniavs': miniavs,
    'notionists': notionists,
    'notionists-neutral': notionistsNeutral,
    'open-peeps': openPeeps,
    'personas': personas,
    'pixel-art': pixelArt,
    'pixel-art-neutral': pixelArtNeutral,
    'rings': rings,
    'shapes': shapes,
    'thumbs': thumbs,
}

/** FIFO cache for generated SVGs (key = `${style}:${seed}`). */
const svgCache = new Map<string, string>()
const MAX_CACHE_SIZE = 500

/**
 * Generate a DiceBear avatar SVG string locally.
 * Results are cached; unknown styles fall back to `thumbs` (Yappr behavior).
 */
export function generateAvatarSvg(style: string, seed: string): string {
    const cacheKey = `${style}:${seed}`
    const cached = svgCache.get(cacheKey)
    if (cached) return cached

    const styleModule = STYLE_MAP[style] || thumbs
    const svg = createAvatar(styleModule, { seed }).toString()

    if (svgCache.size >= MAX_CACHE_SIZE) {
        const firstKey = svgCache.keys().next().value
        if (firstKey) svgCache.delete(firstKey)
    }
    svgCache.set(cacheKey, svg)
    return svg
}

/**
 * Turn a Rust-resolved avatar source into an <img>-bindable src:
 * DiceBear → generated-locally data URI; Uri → the (gateway-resolved) URL.
 */
export function avatarSrc(source: ISocialAvatarSource | undefined | null, identityId: string): string {
    if (source?.kind === 'uri') return source.uri
    const style = source?.kind === 'dicebear' ? source.style : DEFAULT_STYLE
    const seed = source?.kind === 'dicebear' ? source.seed : identityId
    return `data:image/svg+xml;utf8,${encodeURIComponent(generateAvatarSvg(style, seed))}`
}

File: src-tauri/src/social/avatar.rs
// src-tauri/src/social/avatar.rs

//! Yappr profile `avatar` field resolution (port of evonext-mobile
//! `src/helpers/avatar.ts` minus SVG generation, which stays TS-side — see
//! module docs). The field is either:
//!   - a DiceBear JSON string: {"style":"bottts","seed":"xyz"}
//!   - a direct image URI: http(s)://… or ipfs://…
//!
//! Priority (Yappr's resolve-user-details): Yappr `avatar` field →
//! DashPay `avatarUrl` → deterministic DiceBear `thumbs` seeded by the
//! identity ID. Unknown DiceBear styles fall back to `thumbs` at
//! generation time (TS island), matching Yappr.

use super::AvatarSource;

/// Default style for identity-seeded fallback avatars (Yappr uses thumbs).
pub const DEFAULT_STYLE: &str = "thumbs";

/// IPFS gateway for ipfs:// avatar URIs (same default as Yappr).
pub const IPFS_GATEWAY: &str = "https://ipfs.io/ipfs/";

/// Convert an ipfs:// URI to an HTTP gateway URL. Other URIs pass through.
pub fn ipfs_to_gateway(uri: &str) -> String {
    if !uri.starts_with("ipfs://") {
        return uri.to_string();
    }
    let cid_path = uri
        .trim_start_matches("ipfs://")
        .trim_start_matches("ipfs/");
    format!("{IPFS_GATEWAY}{cid_path}")
}

fn is_image_uri(field: &str) -> bool {
    field.starts_with("http://") || field.starts_with("https://") || field.starts_with("ipfs://")
}

/// Parse a Yappr profile `avatar` field into a renderable source.
/// Falls back to a deterministic DiceBear avatar seeded by the identity ID
/// when the field is missing or malformed (Yappr behavior).
pub fn parse_avatar_field(avatar_field: Option<&str>, identity_id: &str) -> AvatarSource {
    if let Some(field) = avatar_field.filter(|f| !f.is_empty()) {
        if is_image_uri(field) {
            return AvatarSource::Uri {
                uri: ipfs_to_gateway(field),
            };
        }
        if let Ok(config) = serde_json::from_str::<serde_json::Value>(field) {
            if let Some(seed) = config.get("seed").and_then(|v| v.as_str()) {
                let style = config
                    .get("style")
                    .and_then(|v| v.as_str())
                    .unwrap_or(DEFAULT_STYLE);
                return AvatarSource::Dicebear {
                    style: style.to_string(),
                    seed: seed.to_string(),
                };
            }
        }
    }
    AvatarSource::Dicebear {
        style: DEFAULT_STYLE.to_string(),
        seed: identity_id.to_string(),
    }
}

/// Resolve the best avatar for a post author.
/// Priority: Yappr profile avatar field → legacy DashPay avatarUrl →
/// deterministic DiceBear from identity ID.
pub fn resolve_author_avatar(
    yappr_avatar_field: Option<&str>,
    dashpay_avatar_url: Option<&str>,
    identity_id: &str,
) -> AvatarSource {
    if let Some(field) = yappr_avatar_field.filter(|f| !f.is_empty()) {
        return parse_avatar_field(Some(field), identity_id);
    }
    if let Some(url) = dashpay_avatar_url.filter(|f| !f.is_empty()) {
        return AvatarSource::Uri {
            uri: ipfs_to_gateway(url),
        };
    }
    parse_avatar_field(None, identity_id)
}
